Consuming openexr through vcpkg breaks CUDA compilation

Commit 03a4718194f68a7056a2119041fccd11230d7007 changed line 13 of `LibraryDefine.cmake`:

```
-    set(_openexr_extra_flags "/EHsc")
+    set(_openexr_extra_flags "/EHsc" "/MP") 
```

and later these flags are added to the `PUBLIC` interface of a target:

```
if (_openexr_extra_flags)
    target_compile_options(${objlib} PUBLIC ${_openexr_extra_flags})
endif()
```

This means that any target depending on openexr through CMake inherits those `PUBLIC` flags.

IMO, it's bad form for openexr to push compile flags like `/EHsc` or `/MP` onto clients of openexr; it's not openexr's job to decide if my code is compiled in parallel or not.

Furthermore, when my target has included CUDA source files, these flags are sent to nvcc to compile the CUDA sources and then the `/MP` flag is unrecognized by nvcc and causes nvcc to issue an error about attempting to provide multiple source files when using `-c` (single file to single object compilation).

To reproduce this issue on Windows (the problem is MSVC specific):

1. `git clone https://github.com/NVIDIA/optix-toolkit`
2. `cd optix-toolkit`
3. `git checkout eaa788db85713ed291d9729a759cb9ef177ff9e8`
4. `git submodule init`
5. `git submodule update`
6. `cd vcpkg`
7. `git pull origin master`
8. `git checkout 7685d93437b0faac15d535f0b6c110be27a95bb8`
9. `cd ..`
10. `cmake --preset otk-default`
11. `cmake --build --preset otk-default --target testDemandLoading`

You get errors similar to the following:

```
  D:\src\tmp\repro\optix-toolkit\build-otk-default\DemandLoading\DemandLoading\tests
  > "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.5\bin\nvcc.exe"  --use-local-env -ccbin "C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.43.34808\bin\HostX64\x64" -x cu   -I"D:\src\tmp\repro\optix-toolkit\DemandLoading\DemandLoading\tests\..\src" -I"D:\src\tmp\repro\optix-toolkit\build-otk-default\DemandLoading\DemandLoading\tests\include" -I"D:\src\tmp\rep
  ro\optix-toolkit\build-otk-default\DemandLoading\DemandLoading\tests\testDemandLoadingKernels" -I"D:\src\tmp\repro\optix-toolkit\build-otk-default\DemandLoading\DemandLoading\DemandLoadingKernels" -I"D:\src\tmp\repro\optix-toolkit\DemandLoading\DemandLoading\include" -I"D:\src\tmp\repro\optix-toolkit\DemandLoading\ImageSource\include" -I"D:\src\tmp\repro\optix-toolkit\build-otk-default\DemandLo
  ading\ImageSource\include" -I"D:\src\tmp\repro\optix-toolkit\ShaderUtil\include" -I"D:\src\tmp\repro\optix-toolkit\Memory\include" -I"C:\ProgramData\NVIDIA Corporation\OptiX SDK 8.0.0\include" -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.5\include" -I"D:\src\tmp\repro\optix-toolkit\build-otk-default\vcpkg_installed\x64-windows\include" -I"D:\src\tmp\repro\optix-toolkit\build-otk-de
  fault\vcpkg_installed\x64-windows\include\OpenEXR" -I"D:\src\tmp\repro\optix-toolkit\build-otk-default\vcpkg_installed\x64-windows\include\Imath" -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.5\include"     --keep-dir testDema.9024B861\x64\Release -use_fast_math -maxrregcount=0   --machine 64 --compile -cudart shared -std=c++17 --generate-code=arch=compute_60,code=[compute_60] /MP -
  Xcompiler="/EHsc -Ob2 /wd4324,/wd4505"   -D_WINDOWS -DNDEBUG -DOTK_USE_CUDA_MEMORY_POOLS -DNOMINMAX -DOPTIX_SAMPLE_USE_CORE_EXR -DOPENEXR_DLL -DIMATH_DLL -DGTEST_LINKED_AS_SHARED_LIBRARY=1 -D"CMAKE_INTDIR=\"Release\"" -D_MBCS -DWIN32 -D_WINDOWS -DNDEBUG -DOTK_USE_CUDA_MEMORY_POOLS -DNOMINMAX -DOPTIX_SAMPLE_USE_CORE_EXR -DOPENEXR_DLL -DIMATH_DLL -DGTEST_LINKED_AS_SHARED_LIBRARY=1 -D"CMAKE_INTDIR
  =\"Release\"" -Xcompiler "/EHsc /W4 /nologo /O2 /FS   /MD " -Xcompiler "/FdtestDemandLoading.dir\Release\vc143.pdb" -o testDemandLoading.dir\Release\/TestSparseVsDenseTextures.cu.obj "D:\src\tmp\repro\optix-toolkit\DemandLoading\DemandLoading\tests\TestSparseVsDenseTextures.cu"
  nvcc fatal   : A single input file is required for a non-link phase when an outputfile is specified
C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Microsoft\VC\v170\BuildCustomizations\CUDA 12.5.targets(799,9): error MSB3721: The command ""C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.5\bin\nvcc.exe"  --use-local-env -ccbin "C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.43.34808\bin\HostX64\x64" -x cu   -I"D:\src\tmp\repro\optix-toolkit
\DemandLoading\DemandLoading\tests\..\src" -I"D:\src\tmp\repro\optix-toolkit\build-otk-default\DemandLoading\DemandLoading\tests\include" -I"D:\src\tmp\repro\optix-toolkit\build-otk-default\DemandLoading\DemandLoading\tests\testDemandLoadingKernels" -I"D:\src\tmp\repro\optix-toolkit\build-otk-default\DemandLoading\DemandLoading\DemandLoadingKernels" -I"D:\src\tmp\repro\optix-toolkit\DemandLoading
\DemandLoading\include" -I"D:\src\tmp\repro\optix-toolkit\DemandLoading\ImageSource\include" -I"D:\src\tmp\repro\optix-toolkit\build-otk-default\DemandLoading\ImageSource\include" -I"D:\src\tmp\repro\optix-toolkit\ShaderUtil\include" -I"D:\src\tmp\repro\optix-toolkit\Memory\include" -I"C:\ProgramData\NVIDIA Corporation\OptiX SDK 8.0.0\include" -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA
\v12.5\include" -I"D:\src\tmp\repro\optix-toolkit\build-otk-default\vcpkg_installed\x64-windows\include" -I"D:\src\tmp\repro\optix-toolkit\build-otk-default\vcpkg_installed\x64-windows\include\OpenEXR" -I"D:\src\tmp\repro\optix-toolkit\build-otk-default\vcpkg_installed\x64-windows\include\Imath" -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.5\include"     --keep-dir testDema.9024B861\
x64\Release -use_fast_math -maxrregcount=0   --machine 64 --compile -cudart shared -std=c++17 --generate-code=arch=compute_60,code=[compute_60] /MP -Xcompiler="/EHsc -Ob2 /wd4324,/wd4505"   -D_WINDOWS -DNDEBUG -DOTK_USE_CUDA_MEMORY_POOLS -DNOMINMAX -DOPTIX_SAMPLE_USE_CORE_EXR -DOPENEXR_DLL -DIMATH_DLL -DGTEST_LINKED_AS_SHARED_LIBRARY=1 -D"CMAKE_INTDIR=\"Release\"" -D_MBCS -DWIN32 -D_WINDOWS -DNDE
BUG -DOTK_USE_CUDA_MEMORY_POOLS -DNOMINMAX -DOPTIX_SAMPLE_USE_CORE_EXR -DOPENEXR_DLL -DIMATH_DLL -DGTEST_LINKED_AS_SHARED_LIBRARY=1 -D"CMAKE_INTDIR=\"Release\"" -Xcompiler "/EHsc /W4 /nologo /O2 /FS   /MD " -Xcompiler "/FdtestDemandLoading.dir\Release\vc143.pdb" -o testDemandLoading.dir\Release\/TestSparseVsDenseTextures.cu.obj "D:\src\tmp\repro\optix-toolkit\DemandLoading\DemandLoading\tests\Tes
tSparseVsDenseTextures.cu"" exited with code 1. [D:\src\tmp\repro\optix-toolkit\build-otk-default\DemandLoading\DemandLoading\tests\testDemandLoading.vcxproj]
```

Applying the attached patch to openexr with a vcpkg overlay port corrects the problem by using a generator expression to only supply the flags when C++ compilation is being performed.  It's still dubious that these are on the `PUBLIC` interface of the openexr target, however.  Certainly the `/MP` flag should be `PRIVATE` for openexr and not `PUBLIC`.

[fix-cuda.patch](https://github.com/user-attachments/files/19652636/fix-cuda.patch)

--- a/cmake/LibraryDefine.cmake
+++ b/cmake/LibraryDefine.cmake
@@ -10,7 +10,7 @@ function(OPENEXR_DEFINE_LIBRARY libname)
   cmake_parse_arguments(OPENEXR_CURLIB "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
 
   if (MSVC)
-    set(_openexr_extra_flags "/EHsc" "/MP")
+    set(_openexr_extra_flags "$<$<COMPILE_LANGUAGE:CXX>:/EHsc>" "$<$<COMPILE_LANGUAGE:CXX>:/MP>")
   endif()
   set(objlib ${libname})
   add_library(${objlib}
@@ -59,7 +59,7 @@ function(OPENEXR_DEFINE_LIBRARY libname)
       target_compile_definitions(${objlib} PUBLIC OPENEXR_USE_DEFAULT_VISIBILITY)
   endif()
   if (_openexr_extra_flags)
-    target_compile_options(${objlib} PUBLIC ${_openexr_extra_flags})
+    target_compile_options(${objlib} PRIVATE ${_openexr_extra_flags})
   endif()
   set_property(TARGET ${objlib} PROPERTY PUBLIC_HEADER ${OPENEXR_CURLIB_HEADERS})
 

